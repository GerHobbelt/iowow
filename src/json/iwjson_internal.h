#pragma once
#ifndef JBL_INTERNAL_H
#define JBL_INTERNAL_H

/**************************************************************************************************
 * IOWOW library
 *
 * MIT License
 *
 * Copyright (c) 2012-2024 Softmotions Ltd <info@softmotions.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *************************************************************************************************/

#include "iwlog.h"
#include "iwjson.h"
#include "iwbinn.h"
#include "iwpool.h"
#include "iwconv.h"

#define JBL_MAX_NESTING_LEVEL 999

struct jbl {
  binn bn;
  struct jbl_node *node;
};

/**
 * @brief struct jbl* visitor context
 */
typedef struct _JBL_VCTX {
  binn *bn;         /**< Root node from which started visitor */
  void *op;         /**< Arbitrary opaque data */
  void *result;
  struct iwpool *pool; /**< Pool placeholder, initialization is responsibility of `JBL_VCTX` creator */
  int  pos;            /**< Aux position, not actually used by visitor core */
  bool terminate;
  bool found;       /**< Used in _jbl_at() */
} JBL_VCTX;

typedef jbn_visitor_cmd_t jbl_visitor_cmd_t;

typedef struct _JBL_PATCHEXT {
  const JBL_PATCH *p;
  JBL_PTR path;
  JBL_PTR from;
} JBL_PATCHEXT;

typedef struct _JBLDRCTX {
  struct iwpool   *pool;
  struct jbl_node *root;
} JBLDRCTX;

iwrc jbl_from_buf_keep_onstack(struct jbl *jbl, void *buf, size_t bufsz);
iwrc jbl_from_buf_keep_onstack2(struct jbl *jbl, void *buf);

iwrc _jbl_write_double(double num, jbl_json_printer pt, void *op);
iwrc _jbl_write_int(int64_t num, jbl_json_printer pt, void *op);
iwrc _jbl_write_json_string(const char *str, int len, jbl_json_printer pt, void *op, jbl_print_flags_t pf);
iwrc _jbl_node_from_binn(const binn *bn, struct jbl_node **node, bool clone_strings, struct iwpool *pool);
iwrc _jbl_binn_from_node(binn *res, struct jbl_node *node);
iwrc _jbl_from_node(struct jbl *jbl, struct jbl_node *node);
bool _jbl_at(struct jbl *jbl, JBL_PTR jp, struct jbl *res);
int _jbl_compare_nodes(struct jbl_node *n1, struct jbl_node *n2, iwrc *rcp);

typedef jbl_visitor_cmd_t (*JBL_VISITOR)(int lvl, binn *bv, const char *key, int idx, JBL_VCTX *vctx, iwrc *rc);
iwrc _jbl_visit(binn_iter *iter, int lvl, JBL_VCTX *vctx, JBL_VISITOR visitor);

bool _jbl_is_eq_atomic_values(struct jbl *v1, struct jbl *v2);
int _jbl_cmp_atomic_values(struct jbl *v1, struct jbl *v2);

BOOL binn_read_next_pair2(int expected_type, binn_iter *iter, int *klidx, char **pkey, binn *value);

#endif
