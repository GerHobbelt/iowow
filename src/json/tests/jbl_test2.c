#include "iwjson.h"
#include "iwjson_internal.h"
#include "iwutils.h"

#include <stdlib.h>
#include <CUnit/Basic.h>

int init_suite(void) {
  int rc = iw_init();
  return rc;
}

int clean_suite(void) {
  return 0;
}

static void test_jbn_xml(void) {
  IWXSTR *xstr = iwxstr_create();
  struct iwpool *pool = iwpool_create_empty();
  CU_ASSERT_PTR_NOT_NULL_FATAL(pool);

  JBL_NODE n;
  const char *val = "{"
                    "\">attr\":\"attrvalue\""
                    "\">attr2\":\"ss < ' xx\""
                    ",\"foo\":\"bar\""
                    ",\"\":\"bo\\ndy 1 < 2 \""
                    ",\"arr\":[1,2,3,{\"obj\":\"k&k\"}]"
                    ",\"nested\":{\"baz\":{\"zaz\":true,\"arr\":[1,3.36,true,\"a\"]}}"
                    "}";
  iwrc rc = jbn_from_json(val, &n, pool);
  CU_ASSERT_EQUAL_FATAL(rc, 0);

  rc = jbn_as_xml(n, &(struct jbn_as_xml_spec) {
    .flags = JBL_PRINT_PRETTY,
    .print_xml_header = true,
    .printer_fn = jbl_xstr_json_printer,
    .printer_fn_data = xstr,
  });
  CU_ASSERT_EQUAL_FATAL(rc, 0);

  char *buf = iwu_file_read_as_buf("data/jbl_test2.001.expected.xml");
  CU_ASSERT_PTR_NOT_NULL_FATAL(buf);
  CU_ASSERT_EQUAL(strcmp(buf, iwxstr_ptr(xstr)), 0);

  iwpool_destroy(pool);
  iwxstr_destroy(xstr);
  free(buf);
}

int main(void) {
  CU_pSuite pSuite = NULL;
  if (CUE_SUCCESS != CU_initialize_registry()) {
    return CU_get_error();
  }
  pSuite = CU_add_suite("jbl_test1", init_suite, clean_suite);
  if (NULL == pSuite) {
    CU_cleanup_registry();
    return CU_get_error();
  }

  if (NULL == CU_add_test(pSuite, "test_jbn_xml", test_jbn_xml)) {
    CU_cleanup_registry();
    return CU_get_error();
  }

  CU_basic_set_mode(CU_BRM_VERBOSE);
  CU_basic_run_tests();
  int ret = CU_get_error() || CU_get_number_of_failures();
  CU_cleanup_registry();
  return ret;
}
