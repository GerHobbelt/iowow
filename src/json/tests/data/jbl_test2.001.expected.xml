<?xml version="1.0" encoding="UTF-8"?>
<root attr="attrvalue" attr2="ss < &#39; xx">
 <foo>bar</foo>
 bo
dy 1 &#60; 2 
 <arr>
  <item>1</item>
  <item>2</item>
  <item>3</item>
  <item>
  <item>
   <obj>k&#38;k</obj>
  </item>
  </item>
 </arr>
 <nested>
  <baz>
   <zaz>true</zaz>
   <arr>
    <item>1</item>
    <item>3.36</item>
    <item>true</item>
    <item>a</item>
   </arr>
  </baz>
 </nested>
</root>